File: informatykaEuropejczyka/Rozdzial2/prog2_36.cpp
//szyfrowanie metoda kolumnowa
#include <iostream>
#include <cstring>
using namespace std;
const int MAX=100;

string szyfruj (string tekst, int klucz[], int n)
{
 string wynik;
 char tablica[MAX][MAX];
 int dl=tekst.size(), m;
 if (dl%n) m=dl/n+1;
 else m=dl/n;
 int k=0;
 for (int i=0;i<m;i++)
  for (int j=0;j<n;j++)
   if (k<dl)
   {
    tablica[i][j]=tekst[k];
    k++;
   }
   else tablica[i][j]='\0';
 int l;
 for (int k=0;k<n;k++)
 {
  l=klucz[k];
  for (int i=0;i<m;i++)
   if (tablica[i][l]!='\0') wynik+=tablica[i][l];
 }
 return wynik;
}

int main()
{
 string tekst;
 int n, klucz[MAX];
 cout<<"podaj tekst jawny: ";
 cin>>tekst;
 cout<<"podaj liczbe kolumn: ";
 cin>>n;
 cout<<"kluczem jest permutacja numerow kolumn z zakresu: 0 - "<<n-1<<endl;
 for (int j=0;j<n;j++)
 {
  cout<<"podaj element klucza nr "<<j+1<<": ";
  cin>>klucz[j];
 }
 cout<<"kryptogram: "<<szyfruj(tekst,klucz,n)<<endl;
 return 0;
}

File: informatykaEuropejczyka/Rozdzial2/prog2_36.py
#szyfrowanie metoda kolumnowa

def szyfruj (tekst, klucz):
    n,dl=len(klucz),len(tekst)
    if dl%n:
        m=dl//n+1
    else:
        m=dl//n
    tablica=[[0 for col in range(n)] for row in range(m)]    
    k=0
    for i in range(m):
        for j in range(n):
            if k<dl:
                tablica[i][j]=tekst[k]
                k+=1
            else:
                tablica[i][j]='\0'
    wynik=''
    for k in range(n):
        l=klucz[k]
        for i in range(m):
            if tablica[i][l]!='\0':
                wynik+=tablica[i][l]
    return wynik

#print(szyfruj('kryptoanaliza',[2,1,4,0,3]))

